R/convert_param.R

```diff
@@ -1,49 +1,301 @@
+#' @param sdlog The sdlog parameter (standard deviation of the natural
+#' logarithm) of the distribution. Also commonly called and denoted by sigma.
+#'
+#' @return A list of eight elements including: mean, median, mode,
+#' variance (`var`), standard deviation (`sd`), coefficient of variation (`cv`),
+#' skewness, and excess kurtosis (`ex_kurtosis`).
+#' @export
+#'
+#' @examples
+#' convert_lnorm_params(meanlog = 1, sdlog = 1)
+convert_lnorm_params <- function(meanlog, sdlog) {
+
+  # check input
+  checkmate::assert_number(meanlog)
+  checkmate::assert_number(sdlog, lower = 0)
+
+  # calculate metrics
+  mean <- exp(meanlog + sdlog^2 / 2)
+  median <- exp(meanlog)
+  mode <- exp(meanlog - sdlog^2)
+  var <- (exp(sdlog^2) - 1) * exp(2 * meanlog + sdlog^2)
+  sd <- sqrt(var)
+  cv <- sd / mean
+  skewness <- (exp(sdlog^2) + 2) * sqrt(exp(sdlog^2) - 1)
+  ex_kurtosis <- exp(4 * sdlog^2) + 2 *
+    exp(3 * sdlog^2) + 3 * exp(2 * sdlog^2) - 6
+
+  # return list of metrics
+  list(
+    mean = mean,
+    median = median,
+    mode = mode,
+    var = var,
+    sd = sd,
+    cv = cv,
+    skewness = skewness,
+    ex_kurtosis = ex_kurtosis
+  )
+}
+
+#' Converts the summary statistics to parameters of the lognormal distribution
+#'
+#' @description Converts the summary statistics input into the meanlog and sdlog
+#' parameters of the lognormal distribution.
+#'
+#' @details The arguments input are captured and handled by name. Therefore,
+#' the names of the arguments must be correct (case-sensitive). Possible input
+#' summary statistics are: `mean`, `median`, `mode`, `var`, `sd`, `cv`,
+#' `skewness`, `ex_kurtosis`.
+#'
+#' Not every combination of input statistics is a viable conversion, and may
+#' error. Common conversions that are implemented are mean and any measure of
+#' spread (`var`, `sd` and `cv`), or `median` and `sd`.
+#'
+#' @param ... Summary statistics to be used for calculating the lognormal
+#' distribution parameters. All arguments must be correctly named, see details
+#' for possible input.
+#'
+#' @return A list of two elements, the meanlog and sdlog
+#' @export
+#'
+#' @examples
+#' convert_lnorm_summary_stats(mean = 3, sd = 2)
+convert_lnorm_summary_stats <- function(...) {
+
+  # capture input
+  x <- list(...)
+
+  # check input
+  chk_ss(x)
+
+  # convert var or cv into sd if available
+  x <- get_sd(x)
+
+  if (is_number(x$mean) && is_number(x$sd)) {
+    # mean and sd to params
+    return(lnorm_meansd2meanlogsdlog(mean = x$mean, sd = x$sd))
+  } else if (is_number(x$median) && is_number(x$sd)) {
+    # median and sd to params
+    sdlog <- sqrt(log(1 + (x$sd / x$median)^2))
+    meanlog <- log(x$median) - sdlog^2 / 2
+  }
+
+  # if either parameter hasn't been calculated error
+  if (!exists("meanlog") || !exists("sdlog")) {
+    stop("Cannot calculate lognormal parameters from given input")
+  }
+
+  # return list of params
+  list(
+    meanlog = meanlog,
+    sdlog = sdlog
+  )
+}
```

Comment:
```suggestion
convert_lnorm_summary_stats <- function(...) {

  # capture input
  x <- list(...)

  # check input
  chk_ss(x)

  # convert var or cv into sd if available
  x <- get_sd(x)

  if (is_number(x$mean) && is_number(x$sd)) {
    # mean and sd to params
    return(lnorm_meansd2meanlogsdlog(mean = x$mean, sd = x$sd))
  }
  if (!(is_number(x$median) && is_number(x$sd))) {
    stop("Cannot calculate lognormal parameters from given input")
  }

  # median and sd to params
  sdlog <- sqrt(log(1 + (x$sd / x$median)^2))
  meanlog <- log(x$median) - sdlog^2 / 2

  # return list of params
  list(
    meanlog = meanlog,
    sdlog = sdlog
  )
}
```

To avoid the `exists()` pattern which is a likely source of error if you rename the variables.